### PR TITLE
[COMCTL32] Fix large combo box pull down closing too quickly

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Improve mouse combo box selection closing immediately._

JIRA issue: [CORE-18769](https://jira.reactos.org/browse/CORE-18769)

This is a retry of PR #4960.

## Proposed changes

_Do not handle CBF_CAPTURE/ReleaseCapture() in COMBO_MouseMove._

### CODE REVIEW

**Comment by HBelusca** [comment]
Note that the standard legacy combobox list code in user32,
https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/user/user32/controls/combo.c;hb=9c1cfe058706ff31dbcd3c7b53eec879f66c8030
also has the same sort of code, thus I have two questions:
- does the corresponding wine's code (in their user32) have the problem too?
- does it too need an adjustment?

**Comment by Doug-Lyons** [comment]
> Note that the standard legacy combobox list code in user32, https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/user/user32/controls/combo.c;hb=9c1cfe058706ff31dbcd3c7b53eec879f66c8030 also has the same sort of code, thus I have two questions:
> 
> * does the corresponding wine's code (in their user32) have the problem too?
> * does it too need an adjustment?

I do not have any specific tests that indicate that this is needed there, but it seems to be the pattern in the past,
so I have added the changes to that file as well. Thanks.

File: win32ss/user/user32/controls/combo.c [hunk 1]
```diff
@@ -2039,6 +2041,10 @@ LRESULT WINAPI ComboWndProc_common( HWND hwnd, UINT message, WPARAM wParam, LPAR
     case WM_MOUSEMOVE:
         if ( lphc->wState & CBF_CAPTURE )
             COMBO_MouseMove( lphc, wParam, lParam );
+#ifdef __REACTOS__
+        if ( lphc->wState & CBF_DROPPED )
+             lphc->wState |= CBF_CAPTURE;
```
Comment (JoachimHenze) [inline win32ss/user/user32/controls/combo.c:2046]:
```suggestion
        if (lphc->wState & CBF_DROPPED)
            lphc->wState |= CBF_CAPTURE;
```

File: dll/win32/comctl32/combo.c [hunk 2]
```diff
@@ -1610,8 +1610,9 @@ static void COMBO_LButtonDown( LPHEADCOMBO lphc, LPARAM lParam )
        else
        {
 	   /* drop down the listbox and start tracking */
-
+#ifndef __REACTOS__
            lphc->wState |= CBF_CAPTURE;
+#endif
```
Comment (HBelusca) [inline dll/win32/comctl32/combo.c:1615]: I would like to understand the following.
This code path still starts the capture (see below); however now the flag that says "capture" is not set, so there's some sort of "discrepancy" between the actual state and the translation in terms of flags (in `wState`).

Is the capture started by `SetCapture()` correctly reset later, whatever the value of `wState` ?

Comment (Doug-Lyons) [reply dll/win32/comctl32/combo.c]: @HBelusca, Please review after my recent changes and see if the wState doesn't track the SetCapture() now while still retaining the improvements to the drop-down listbox. Thanks.

**Comment by Doug-Lyons** [comment]
Testbot Results:

KVM:  https://reactos.org/testman/compare.php?ids=93877,93879,93883,93886,93889 LGTM
VBox: https://reactos.org/testman/compare.php?ids=93868,93870,93872,93885,93888 LGTM

There are some variations in comctl32:listbox, but these have been occurring for some time.
That is the reason that I showed more past tests to show this fact.

File: win32ss/user/user32/controls/combo.c (hunk 1 above)
Comment (JoachimHenze) [reply win32ss/user/user32/controls/combo.c]: resolved.

**Comment by JoachimHenze** [comment]
I tested todays artifact
[reactos-clang-i386-Debug-0x502-a50bacd208a448c8a902de73ef0b83c2223e3328](https://github.com/reactos/reactos/actions/runs/8212256028/artifacts/1311312824)
and found a bug:
There are cases when the triangle remains in pressed state erroneously:
**Reproduction steps**
- Expand a combox
- hit Arrow-down 3 times
- hit Arrow-up 3 times
- close the Combobox with the return-key
[BUG_triangle_remains_in_pressed_state.webm](https://github.com/reactos/reactos/assets/33393466/41e968e2-f276-47b2-b680-ba35a45365ac)
Sidenote: The bug will not happen if I close the Combobox with the mouse-button
Sidenote: I have not tested unmodified master head yet, whether this bug happens also without this PR.

**Comment by HBelusca** [comment]
@Doug-Lyons OK now with your latest change the flag is set and reset together with the Set/ReleaseCapture, so it would look more correct to me. Now, regarding what @JoachimHenze just have found, I wonder whether the problem he has found (with the triangle button) always happened, even with the previous iteration of your PR (i.e. before your last flag change I suggested), or not.

**Comment by Doug-Lyons** [comment]
refs/pull/5566/merge
Current Testbot results:

KVM:  https://reactos.org/testman/compare.php?ids=93978,93980
VBox: https://reactos.org/testman/compare.php?ids=93976,93979

**Comment by Doug-Lyons** [comment]
> @Doug-Lyons OK now with your latest change the flag is set and reset together with the Set/ReleaseCapture, so it would look more correct to me. Now, regarding what @JoachimHenze just have found, I wonder whether the problem he has found (with the triangle button) always happened, even with the previous iteration of your PR (i.e. before your last flag change I suggested), or not.

@HBelusca, I have not been able to fix the problem pointed out by @JoachimHenze and looked more closely at the way the present code works. Now I will try a totally different approach and area of the code to modify. Let's see if it works OK first from @JoachimHenze before continuing. Thanks.

**Comment by JoachimHenze** [comment]
I retested todays PR state by using artifact https://github.com/reactos/reactos/actions/runs/8239700313/artifacts/1316502291
And I have to say, I am very impressed. This works **really good** already, definitely better than the master head state.

But nevertheless I found a small glitch, if we could fix that as well before the merge, that would be absolutely perfect:
- use the timedate_unlimited.cpl from the ticket
- switch to 640x480 and place timedate in the middle of y
- look closely what the current selection is in the combobox (for me it is Berlin for example)
- now click to open the combobox (it will struggle as intended to make it fit the screen, it will open without autoselection, good)
- but you will now see that while it is opened, the current selection is not Berlin any longer, but some other entry from the list
- if you cancel now via ESC key, the unintended new selection will be chosen in the collapsed combobox
[video-2024-03-29T19-47-46-046875000Z.webm](https://github.com/reactos/reactos/assets/33393466/c6389951-03d4-446a-8b31-b3c861a39b7f)
everytime the combobox collapses in the video, I used the ESCAPE-key, not a click, and not enter.
The area where the mouse is over should not be taken then.

Contrary on Windows it will stick with the "Berlin"-selection after pressing Escape here.

That is really nitpicking already. But if we do touch this (also with the aim to propose to Wine) that stuff should work flawlessly.
I do therefore recommend another iteration of improvement before merging this.
**But like I expressed already: I do definitely think you do outperfom the master-head-state already in this context!**

**Comment by HBelusca** [comment]
@JoachimHenze Regarding this:
> Contrary on Windows it will stick with the "Berlin"-selection after pressing Escape here.

does Windows also stay on the "Berlin"-selection (contrary to ReactOS/Wine's combobox) while it's doing this:
> [...] while it is opened, the current selection is not Berlin any longer, but some other entry from the list

?

**Comment by JoachimHenze** [comment]
@hbelusca it is hard to describe it without a video, but I will try.
When you do the click on XPSP3 on the text in the combobox, it opens the combobox and there are 2 markings in there.
One in blue, which will stick with the "Berlin" selection initially if you hold the mouse absolutely still.
There is a 2nd dotted selection in that combobox, which immediately is marking an arbitrary element in the combobox that is beneath your mouse cursor.

If you move the mouse now a slightest bit to the top or to the bottom more than the current elements height, then the **blue** selection will change from Berlin to the (then new) element under your cursor.

But whatever the **blue** selection will be at this point, if you press ESCAPE, it will **always** stick with the selection that was active when you opened the combobox: "Berlin" in my case.

**Comment by HBelusca** [comment]
@Doug-Lyons Patch now looks simpler (you just disable pieces of existing wine code only).
Also @JoachimHenze : Is this last PR iteration OK or still has the problem you've described? Is current master (w/o PR) also affected?

**Comment by Doug-Lyons** [comment]
Closing now because this has been superseded by PR https://github.com/reactos/reactos/pull/6789 by @DarkFire01.
Note that there are still unresolved issues with: [CORE-18769](https://jira.reactos.org/browse/CORE-18769).

**Pull request closed without merging**